src/controller.cpp

```diff
@@ -0,0 +1,207 @@
+#include <memory>
+#include <iostream>
+#include <sstream>
+#include <fstream>
+#include <set>
+
+#include "Controller.h"
+#include "Board.h"
+
+using Chess::Board;
+using Chess::Controller;
+using Chess::utilities::Color;
+using Chess::utilities::PlayerType;
+
+static const char *BRIGHT = "\033[1m";
+static const char *RESET = "\033[0m";
+static const char *RED_FG = "\033[31m";
+static const char *BLUE_FG = "\033[34m";
+
+Controller::Controller(std::string mode) : white(nullptr), black(nullptr), board(nullptr)
+{
+	init(mode);
+}
+
+void Controller::init(const std::string &type)
+{
+	if (type.compare("pc") == 0)
+	{
+		srand(time(NULL));
+		bool white_is_human = rand() % 2 == 1;
+
+		white = std::make_shared<Player>(Color::white, PlayerType{white_is_human ? PlayerType::human : PlayerType::computer});
+		black = std::make_shared<Player>(Color::black, PlayerType{white_is_human ? PlayerType::computer : PlayerType::human});
+	}
+	else if (type.compare("cc") == 0)
+	{
+		white = std::make_shared<Player>(Color::white, PlayerType::computer);
+		black = std::make_shared<Player>(Color::black, PlayerType::computer);
+	}
+	else
+	{
+		throw "Invalid game type";
+	}
+
+	std::string base_fen = "rnbqkbnr/pppppppp/8/8/8/8/PPPPPPPP/RNBQKBNR";
+	std::string fen = "1r3k2/6br/n3Q3/pp2Bp1p/2p4P/2N1P3/PPP2PR1/2K5";
+	board = std::make_unique<Board>(base_fen, *white.get(), *black.get());
+}
+
+std::map<char, unsigned int> col_mapping{
+	 {'A', 0},
+	 {'B', 1},
+	 {'C', 2},
+	 {'D', 3},
+	 {'E', 4},
+	 {'F', 5},
+	 {'G', 6},
+	 {'H', 7}};
+
+std::map<char, unsigned int> row_mapping{
+	 {'1', 7},
+	 {'2', 6},
+	 {'3', 5},
+	 {'4', 4},
+	 {'5', 3},
+	 {'6', 2},
+	 {'7', 1},
+	 {'8', 0}};
+
+Chess::Movement Controller::get_move()
+{
+	std::string from = "";
+	std::string to = "";
+
+	std::cin >> from >> to;
+	std::cout << "\033[14A\033[J";
+
+	if (from.compare("XX") == 0 && to.compare("XX") == 0)
```

Comment:
questa implementazione la lasciamo così(secondo me è migliore di quello che chiede il prof, perchè credo che secondo lui fosse più facile XX XX) o la modifichiamo alla fine del progetto dopo tutti i test?

Reply:
No, lasciamo così. Il comando ora come ora stampa una board sopra quella corrente e se si continua a farlo si ottiene una cronologia delle mosse. Altrimenti viene riscritto tutto.
è un modo per tenere traccia delle mosse fatte mentre giochi se vogliamo.

Reply:
L'anno scorso eravamo solo in due gruppi ad aver sperimentato con il controllo del cursore e dei colori con [codici ANSI/VT100](https://web.archive.org/web/20200413124215/http://www.termsys.demon.co.uk/vtansi.htm) di controllo del terminale. Penso sarà lo stesso quest'anno. Il prof non si aspetta una cosa così e generalmente chiede cose più terra terra.

Reply:
Anche l'uso delle lambda va oltre allo scope del progetto, però sono comode.